File: baekjoon/너비_우선_탐색/1707-이분_그래프.java
ASDASDASDimport java.util.*;

public class Main {

    private static final int EMPTY = 0;
    private static final int RED = 1;
    private static final int BLUE = 2;
    private static Scanner sc = new Scanner(System.in);
    private static int k;
    private static int v;
    private static int e;
    private static class Node {
        int number;
        int color;

        public Node(int number, int color) {
            this.number = number;
            this.color = color;
        }
    }

    public static void main(String[] args) {
        k = sc.nextInt();
        for (int i = 0; i < k; i++) {
            if (solution()) System.out.println("YES");
            else System.out.println("NO");
        }
    }

    private static boolean solution() {
        v = sc.nextInt();
        e = sc.nextInt();

        ArrayList<Integer>[] graph = new ArrayList[v+1];
        for (int i = 1; i <= v; i++)
            graph[i] = new ArrayList<>();
        int[] color = new int[v+1];

        fillGraph(graph);

        for (int start = 1; start <= v; start++) {
            if (color[start] == EMPTY) {
                if(!bfs(graph, color, start))
                    return false;
            }
        }

        return true;
    }

    private static void fillGraph(ArrayList<Integer>[] graph) {
        for (int i = 0; i < e; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();

            graph[a].add(b);
            graph[b].add(a);
        }
    }

    private static boolean bfs(ArrayList<Integer>[] graph, int[] color, int start) {
        Queue<Node> q = new LinkedList<>();

        q.add(new Node(start, RED));
        color[start] = RED;

        while (!q.isEmpty()) {
            int qSize = q.size();
            for (int i = 0; i < qSize; i++) {
                Node now = q.poll();

                for (int next : graph[now.number]) {
                    if (color[next] == now.color)
                        return false;

                    if (color[next] == EMPTY) {
                        color[next] = now.color == RED ? BLUE : RED;
                        q.add(new Node(next, color[next]));
                    }
                }
            }
        }

        return true;
    }
}

File: baekjoon/너비_우선_탐색/1707-이분_그래프.py
from collections import deque

EMPTY = 0
RED = 1
BLUE = 2

def bfs(graph, color, start):
    q = deque()
    q.append([start, RED])
    color[start] = RED

    while q:
        now = q.popleft()
        for nxt in graph[now[0]]:
            if color[nxt] == now[1]:
                return False
            
            if color[nxt] == EMPTY:
                color[nxt] = BLUE if now[1] == RED else RED
                q.append([nxt, color[nxt]])

    return True

def solution():
    v, e = map(int, input().split())
    graph = [[] for _ in range(v+1)]
    color = [EMPTY for _ in range(v+1)]

    for _ in range(e):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    for i in range(1, v+1):
        if color[i] == EMPTY:
            if not bfs(graph, color, i):
                return False

    return True

k = int(input())
for _ in range(k):
    if solution():
        print("YES")
    else:
        print("NO")